staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -768,6 +770,120 @@ func (r *Request) Watch(ctx context.Context) (watch.Interface, error) {
 	}
 }
 
+type WatchListResult struct {
+	// err holds any errors we might have received
+	// during streaming.
+	err error
+
+	// store holds collected data
+	store store
+
+	// initialEventsEndBookmarkRV holds the resource version
+	// extracted from the bookmark event that marks
+	// the end of the stream.
+	initialEventsEndBookmarkRV string
+}
+
+func (r WatchListResult) Into(obj runtime.Object) error {
+	if r.err != nil {
+		return r.err
+	}
+
+	listPtr, err := meta.GetItemsPtr(obj)
+	if err != nil {
+		return err
+	}
+	listVal, err := conversion.EnforcePtr(listPtr)
+	if err != nil {
+		return err
+	}
+	if listVal.Kind() != reflect.Slice {
+		return fmt.Errorf("need a pointer to slice, got %v", listVal.Kind())
+	}
+
+	listObjects := r.store.List()
+	if len(listObjects) == 0 {
+		listVal.Set(reflect.MakeSlice(listVal.Type(), 0, 0))
+	} else {
+		listVal.Set(reflect.MakeSlice(listVal.Type(), len(listObjects), len(listObjects)))
+		for i, o := range listObjects {
+			listVal.Index(i).Set(reflect.ValueOf(o).Elem())
+		}
+	}
+
+	listMeta, err := meta.ListAccessor(obj)
+	if err != nil {
+		return err
+	}
+	listMeta.SetResourceVersion(r.initialEventsEndBookmarkRV)
+
+	// TODO: set kind and apiversion
+	return nil
+}
+
+func (r *Request) WatchList(ctx context.Context) WatchListResult {
+	// assume the opts have been properly set
+	w, err := r.Watch(ctx)
+	if err != nil {
+		return WatchListResult{err: err}
+	}
+
+	return r.handleWatchList(ctx, w)
+}
+
+func (r *Request) handleWatchList(ctx context.Context, w watch.Interface) WatchListResult {
+	defer w.Stop()
+	s := newOrderedStore()
+	initialEventsEndBookmarkRV := ""
+
+	for {
+		select {
+		case <-ctx.Done():
+			// TODO: handle
+			//return errorStopRequested
+		case event, ok := <-w.ResultChan():
+			if !ok {
+				// TODO: handle
+			}
+			if event.Type == watch.Error {
+				// TODO: handle error
+				//return apierrors.FromObject(event.Object)
```

Comment:
when handling, it's likely useful to indicate how many items we actually received in the error.